### PR TITLE
Apply templating on alert notifications on OK state

### PR DESCRIPTION
**Pull request by santihernandezc**
**What this PR does / why we need it**:

Alerts that go from `Firing` to `OK` don't have evaluation data associated. This results in templates not being parsed in notification titles and messages.

Screen Shot 2022-04-05 at 15 35 20

With this fix the alerting system checks if there are time series that violate the conditions the user set for the alert rule, and if there are none, it uses the evaluation data from the query.

Screen Shot 2022-04-05 at 15 36 48

**Which issue(s) this PR fixes**:

https://github.com/grafana/grafana/issues/30606

**Special notes for your reviewer**:

We were using the `EvaMatches` field inside the `EvalContext`,  but when the the alert rule is `OK` this field is empty, so there's no data available to use to parse the notification message/title templates.

~Now, the `EvalContext` uses the new `AllSeries` field, which is added when the rule is evaluated.~

Update: after some discussion, we decided that using previous evaluation data was not the right thing to do. This fix uses current evaluation data, using the data source as the single source of truth instead of relying on the database to get the needed data.

Note that this PR only affects the Legacy Alerting system.

### CODE REVIEW

**Comment by alexweav** [comment]
I noticed this fix is only in the legacy alerting directory. Does this problem also happen in ngalerting, or is it specific to legacy? I don't think we support legacy alerting any more, aside from really critical issues, because it's getting deleted entirely in a few months anyway.

**Comment by alexweav** [comment]
I am not sure who all on the team was an expert when legacy alerting was being actively maintained, but I definitely don't feel comfortable hitting approve, since I joined well after we started moving to the new-gen alerting. I don't understand the context/architecture of the old alerting system well enough.

@gotjosh what's our support level for legacy? Who would have enough context to review such a change?

**Comment by santihernandezc** [comment]
@alexweav this PR is meant to solve [issue #30606](https://github.com/grafana/grafana/issues/30606), so it's indeed related to legacy alerting.

**Comment by achatterjee-grafana** [comment]
Should we plan for doc updates?

**Comment by gotjosh** [comment]
Please add a changelog entry. If your PR is backported, there's no changelog in the 8.4.x branch and you'll have to remove the changelog in the backport.

**This is a copy & paste message**

**Comment by JohnnyQQQQ** [comment]
I see I got assigned to this, I have not much context on the old alerting, didn't work on it once. Would be nice if someone else could review this. On another note, the drone build is failing.

**Comment by gotjosh** [comment]
> @gotjosh the PR is not backported, but this fix is only related to legacy alerting.
Nothing inside the UA directories has been modified. Do you think we should add a changelog entry anyway?

Yes, we should!

**Comment by achatterjee-grafana** [comment]
> @achatterjee-grafana I don't think we should, since [docs](https://grafana.com/docs/grafana/latest/alerting/old-alerting/add-notification-template/) currently say:
> 
> > Labels that exist from the evaluation of the alert query can be used in the alert rule name and in the alert notification message fields. The alert label data is injected into the notification fields when the alert is in the alerting state. When there are multiple unique values for the same label, the values are comma-separated.
> 
> That still holds true. If the alert never went into `Alerting` state, then there are no values to parse in the templates. This is not a new feature but a bug fix.

Thanks for the clarification.

File: pkg/services/alerting/rule.go
Comment (gotjosh) [inline pkg/services/alerting/rule.go:212]: Can you give me a brief explanation of what's happening here? In which cases would `EvalData` not be empty and we are we setting  `EvalMatches`?

Comment (santihernandezc) [reply pkg/services/alerting/rule.go:212]: - This field is of type `*simplejson.Json`. Since it's a pointer, calling a method like `.Get()` could result in panics. I've run into this while trying to update alerts, thus the check to see whether it's nil.
- If it's non-nil but empty, then the `.MustArray()` call will return an empty slice, so the range loop won't execute. This happens when the alert rule goes from `Alerting` to `OK` and the database gets updated with empty evaluation matches.
- If it's non-nil and not empty, like every time we fetch all alerts from the database, then the eval matches get translated from `simplejson.Json` to `[]*EvalMatch`.

So, every time we query the database we get back evaluation data. We were not using this before, but now we are parsing that JSON into a slice of the correct type to be used later in the `EvalContext`.

**Comment by santihernandezc** [comment]
After discussing the approach to solve the [original issue](https://github.com/grafana/grafana/issues/30606) with @gotjosh we've decided not to use previous evaluation data, but instead try to pull data from the evaluation engine and use that to parse the templates.
If this approach turns out to be a better alternative, this PR will be refactored focusing on that evaluation data.

File: pkg/services/alerting/eval_context.go
Comment (gotjosh) [inline pkg/services/alerting/eval_context.go:198]: After our offline discussion, I don't think this is correct.

What this section is doing is: Use all the matches if you don't have any alerting matches (given `EvalMatches` only contain those values that would be alerting).

Let's think about a scenario for a second: Let's say you have an alert that observes 5 series

1. All 5 series go into alert, you get an alert that says: "ALERTING s1,s2,s3,s4,s5"
2. On the next iteration 2 them resolve, what you need is two notifications: "ALERTING s1,s2,s3" and "OK s4,s5"
3. On the next iteration all of them resolve "OK s1,s2,s3" 

How does this map to this understanding? In my eyes (with my limited knowledge about how templating works in legacy) what you need is to compute the difference between `EvalMatches` and `AllMatches` and assume those are the non-firing alerts.

Comment (gotjosh) [reply pkg/services/alerting/eval_context.go:198]: Once we figure the above, I'd recommend you scope this out to a function so that you can do something like:

```golang
func (c *EvalContext) GetTemplateMatches() []*EvalMatch {
	// EvalMatches represent alerting values, if we have any this means this is a firing alert and we should use those the data for the template.
	if len(c.EvalMatches) > 0 {
		return c.EvalMatches
	}

	// If we don't have any alerting values, use all of the values to template the alert.
	return c.AllMatches
}
```

**Review by gotjosh — CHANGES_REQUESTED** [review]
Please see my comment about the templating to validate my understanding.

File: pkg/services/alerting/eval_context.go
Comment (santihernandezc) [reply pkg/services/alerting/eval_context.go:198]: Making that diff between `EvalMatches` and `AllMatches` would result in a set of series that didn't violate the constraint for the alert to fire. In that case it would be possible to send an extra `OK` notification.

For item 3 we would need the previous evaluation data. Without that we can't know what which series were firing and got resolved.

Comment (armandgrillet) [reply pkg/services/alerting/eval_context.go:198]: With this PR as it is right now, what would happen at step 2? Nothing? Or will it say "OK s1,s2,s3, s4, s5" despite s1, s2, and s3 still being alerting?

Comment (santihernandezc) [reply pkg/services/alerting/eval_context.go:198]: It would say "ALERTING s1,s2,s3", but we have nothing in place to send an extra OK notification for the series that are no longer firing.
I reproduced this scenario. This is what happens if the notifier is configured to send reminders:

Screen Shot 2022-04-13 at 09 06 25

Two series were causing the alert to fire, then just one, and then all series were OK.
The alert rule name is `Prueba legacy alert: ${instance}`.

**Comment by santihernandezc** [comment]
@achatterjee-grafana now we do need to plan for doc updates:

> The alert label data is injected into the notification fields when the alert is in the alerting state

The data is injected on each evaluation of the alert rule, not only when it is alerting. Otherwise, we wouldn't be able to parse templates in OK state.

We also have to make clear that this is the expected behavior:
- FIRING: the data used to parse the templates correspond to the series that are violating the alert condition at the moment we send the notification.
- OK: all series are used since there are none that are violating the condition.

File: pkg/services/ngalert/CHANGELOG.md
Comment (gotjosh) [inline pkg/services/ngalert/CHANGELOG.md:49]:
```suggestion
- [BUGFIX] (Legacy) Templates: Parse notification templates using all the matches of the alert rule when going from `Alerting` to `OK` #47355
```

**Review by gotjosh — COMMENTED** [review]
LGTM.

Would love to see the docs update before I click that approve button, feel free to ping me again.

**Review by gotjosh — APPROVED** [review]
LGTM. Great job 👏

File: docs/sources/alerting/old-alerting/add-notification-template.md
Comment (achatterjee-grafana) [inline docs/sources/alerting/old-alerting/add-notification-template.md:14]:
```suggestion
You can use labels generated during an alerting query evaluation to create alert notification messages. For multiple unique values for the same label, the values are comma-separated.
				
When an alert fires, the alerting data series indicates the violation. For resolved alerts, all data series are included in the resolved notification. 
```

**Review by achatterjee-grafana — COMMENTED** [review]
Added doc review.

File: docs/sources/alerting/old-alerting/add-notification-template.md
Comment (santihernandezc) [reply docs/sources/alerting/old-alerting/add-notification-template.md:14]: I just created a separate PR for this change: https://github.com/grafana/grafana/pull/47744

**Pull request merged**
